develop/tutorials/articles/160-search/03-liferay-search-api/03-indexers.markdown

```diff
@@ -1,12 +1,195 @@
-## Indexers [](id=indexers)
+# Indexing Framework
```

Comment:
@lipusz I think this tutorial on Indexing is ready for feedback. What additional information should we cover?

Reply:
**Started reviewing...** :mag:

Reply:
>Each asset in @product@ must be indexed in the search engine

It's not a must, unless you want to take advantage of "full-text search" capabilities.

Permission-aware: I'd add a header for this so it can be grabbed and linked easily because it's an important thing.

>Along with the search engine, @product@ has its own search infrastructure. @product@ adds the following features to the existing Elasticsearch API:

I'd wrote something like:
_The heart of Liferay's Search API was originally built around Lucene so if you are familiar with Lucene it can ease understanding Liferay. Starting with 7.0 version of the @product@, Liferay started to re-work the Search API by first generalizing those parts which were closely tied to Lucene and then Elasticsearch support was added besides the Solr support. Since that point, most of the searching and indexing related APIs were developed primary to leverage/map Elasticsearch APIs which means that in certain cases (for example the Query types) there is a one-to-one mapping between a Liferay and an Elasticsearch API._

_Besides these, Liferay' Search Infrastructure includes additional APIs for various purposes such as:... (bullet points come)_

>ModelDocumentContributor classes specify which database fields are indexed in the model entity's search engine documents.

The last word "documents" should be singular I think as 1 entity has 1 index document in the search engine.

Reply:
I know Rich does not prefer it, tough I'd still like to see links to GitHub to the actual class, method. This is for developers and dev docs are usually refer to actual "code". You could you use https://github.com/liferay/liferay-portal/tree/7.1.0-ga1 or the latest "ga".

Without "context" it's not that informative. People like to explore and navigate to see the "actual thing".

Reply:
`ModelIndexerWriterContributor`: it's also invoked each time a CRUD operation on an entity is made if the method `modelIndexed` is implemented in your custom contributor.

Reply:
>To influence the way model entity fields are indexed in search engine documents,

Should be a header (level 3 or 4, whatever makes sense).

> To exert control over the way your model entity documents are searched,

Should be a header (level 3 or 4, whatever makes sense).

Reply:
`KeywordQueryContributor`, `ModelPreFilterContributors`: Add links to GitHub.

Reply:
`KeywordQueryContributor`: Also cover https://grow.liferay.com/learn/Composite+Search+Indexer+Implementation+in+7.1#L5.x_How-to-contribute-certain-query-clauses-to-every-search-in-7.1-regardless-of-what-base-entity-is-being-searched (ignore if it's done later)

Reply:
`ModelPreFilterContributors`: https://grow.liferay.com/learn/Composite+Search+Indexer+Implementation+in+7.1#L5.x_How-to-contribute-certain-Filter-clauses-to-every-search-regardless-of-what-base-entity-is-being-searched and https://grow.liferay.com/learn/Composite+Search+Indexer+Implementation+in+7.1#L5.x_What-is-the-difference-between-QueryPreFilterContributor-and-ModelPreFilterContributor (ignore if it's covered later)

Reply:
>To influence the result summaries for your model entity documents

Should be a header (level 3 or 4, whatever makes sense).

Reply:
`ModelSummaryContributor, ModelVisibilityContributor`: Add links to GitHub.

Reply:
`ModelSummaryContributor`: Registering as `@Component` without an `indexer.class.name` property has the same effect than for KeywordQueryContributor and DocumentContributor.

Reply:
> Registration

Should be a header (level 3 or 4, whatever makes sense).

Reply:
`SearchRegistrar`, `ModelSearchDefinition`: Add links to GitHub.

Reply:
>Registration occurs as soon as the Component is activated (during portal startup).

During portal startup or later when deploying the bundle.

Reply:
>For example, when a Blogs Entry 

If we refer to the entity than it should be written as `BlogsEntry`. Otherwise it's a "Blogs entry".

Reply:
>Service Layer Indexing Requirements

It's more like an "Trigger Indexing from Service Layer through Annotations" or something like that. What do you think?

Reply:
Kudos for the mapping table! :+1: :trophy:

Reply:
> I'd wrote something like:
> _The heart of Liferay's Search API was originally built around Lucene so if you are familiar with Lucene it can ease understanding Liferay. Starting with 7.0 version of the @Product@, Liferay started to re-work the Search API by first generalizing those parts which were closely tied to Lucene and then Elasticsearch support was added besides the Solr support. Since that point, most of the searching and indexing related APIs were developed primary to leverage/map Elasticsearch APIs which means that in certain cases (for example the Query types) there is a one-to-one mapping between a Liferay and an Elasticsearch API._
> 
> _Besides these, Liferay' Search Infrastructure includes additional APIs for various purposes such as:... (bullet points come)_

How about this:

Liferay's original Search API was built around the Lucene search and indexing
library. To this day, familiarity with Lucene will jump-start your understanding
of Liferay's Search API. However, starting with the 7.0 version of @product@,
the Search API is being reworked, so that the parts closely tied to Lucene are
becoming more generic. Elasticsearch support was added (in addition to Solr),
and most of the newer searching and indexing APIs aim to leverage/map
Elasticsearch APIs. This means that in many cases (for example the `Query`
types) there is a one-to-one mapping between the Liferay and Elasticsearch APIs.

In addition to the Elasticsearch centered APIs, Liferay's Search Infrastructure
includes additional APIs serving these purposes: [bullet list follows]